graphs/dijkstra_algorithm.py

```diff
@@ -1,211 +1,520 @@
         self.array[j] = temp
 
-    def decrease_key(self, tup, new_d):
-        idx = self.pos[tup[1]]
-        # assuming the new_d is atmost old_d
+    def decrease_key(self, tup, new_d) -> None:
+        """
+        Decrease the key value for a given tuple, assuming the new_d is at most old_d.
+        Examples:
+        >>> priority_queue_test = PriorityQueue()
+        >>> priority_queue_test.array = [(10, 'A'), (15, 'B')]
+        >>> priority_queue_test.cur_size = len(priority_queue_test.array)
+        >>> priority_queue_test.node_positions = {'A': 0, 'B': 1}
+        >>> priority_queue_test.decrease_key((10, 'A'), 5)
+        >>> priority_queue_test.array
+        [(5, 'A'), (15, 'B')]
+        """
+        idx = self.node_positions[tup[1]]
         self.array[idx] = (new_d, tup[1])
-        while idx > 0 and self.array[self.par(idx)][0] > self.array[idx][0]:
-            self.swap(idx, self.par(idx))
-            idx = self.par(idx)
+        while idx > 0 and self.array[self.parent_idx(idx)][0] > self.array[idx][0]:
+            self.swap(idx, self.parent_idx(idx))
+            idx = self.parent_idx(idx)
 
 
 class Graph:
-    def __init__(self, num):
-        self.adjList = {}  # To store graph: u -> (v,w)
-        self.num_nodes = num  # Number of nodes in graph
-        # To store the distance from source vertex
-        self.dist = [0] * self.num_nodes
-        self.par = [-1] * self.num_nodes  # To store the path
-
-    def add_edge(self, u, v, w):
-        #  Edge going from node u to v and v to u with weight w
-        # u (w)-> v, v (w) -> u
-        # Check if u already in graph
-        if u in self.adjList:
-            self.adjList[u].append((v, w))
-        else:
-            self.adjList[u] = [(v, w)]
-
-        # Assuming undirected graph
-        if v in self.adjList:
-            self.adjList[v].append((u, w))
-        else:
-            self.adjList[v] = [(u, w)]
-
-    def show_graph(self):
-        # u -> v(w)
-        for u in self.adjList:
-            print(u, "->", " -> ".join(str(f"{v}({w})") for v, w in self.adjList[u]))
-
-    def dijkstra(self, src):
-        # Flush old junk values in par[]
-        self.par = [-1] * self.num_nodes
-        # src is the source node
-        self.dist[src] = 0
-        q = PriorityQueue()
-        q.insert((0, src))  # (dist from src, node)
-        for u in self.adjList:
-            if u != src:
-                self.dist[u] = sys.maxsize  # Infinity
-                self.par[u] = -1
-
-        while not q.is_empty():
-            u = q.extract_min()  # Returns node with the min dist from source
-            # Update the distance of all the neighbours of u and
-            # if their prev dist was INFINITY then push them in Q
-            for v, w in self.adjList[u]:
-                new_dist = self.dist[u] + w
-                if self.dist[v] > new_dist:
-                    if self.dist[v] == sys.maxsize:
-                        q.insert((new_dist, v))
```

Comment:
Please revert this change.

Reply:
You have two choices:
1. Leave this pull request as it is and save it for later.  Create a new PR that ONLY contains the doctests.  We will merge it and then come back to this PR.
-- or --
2. Remove all the code changes and keep just the doctests.  We can merge this one and then you can recreate the code changes.

The FIRST is probably safer.